### PR TITLE
Suppress error reporting on unittest.SkipTest

### LINKED ISSUE #514: Hypothesis doesn't recognize unittest.SkipTest exception
# Narrative example of failure

I haven't had time to construct a minimal example, so I'll give a full example of failure. I've recently updated a test in my Django project which conditionally skips a test in cases which no longer make sense. 

The test in question is defined in an abstract base class of a couple other test classes which specialize it further. One branch of one test no longer makes sense in the context of one of the specialization classes. I can't solve this with `assume()` or `@unittest.skipIf()` because neither has enough contextual data. Instead, I `raise unittest.SkipTest()` at runtime if the test should be skipped.

The issue is that Hypothesis doesn't know about `unittest.SkipTest`, so produces failure output, as seen below:

```shell
(.venv) coriolinus$ ./manage.py test
Creating test database for alias 'default'...
Installed 39 object(s) from 1 fixture(s)
Installed 21 object(s) from 1 fixture(s)
Installed 9579 object(s) from 1 fixture(s)
Installed 5 object(s) from 1 fixture(s)
Installed 6 object(s) from 1 fixture(s)
.............................s.............................Falsifying example: test_admin_sees_everyone(self=<core.tests.test_user_profile_visibility.ProfileVisibilityTests testMethod=test_admin_sees_everyone>, data=data(...))
Draw 1: <User: alpha>
s..........................................
----------------------------------------------------------------------
Ran 102 tests in 58.985s

OK (skipped=2)
Destroying test database for alias 'default'...
(.venv) coriolinus$ pip list | grep hypothesis
hypothesis (3.6.1)
```

# Expected behavior

Hypothesis does not produce failure output if a test raises `unittest.SkipTest`.

# Observed behavior

Hypothesis produces failure output if a test raises `unittest.SkipTest`.

### PR DESCRIPTION
**Pull request by alexwlchan**
Resolves #514. Test case per the example in that thread.

Not ready to merge – I’ll need to write some tests for Hypothesis CI, but that will need to wait until I get home this evening.

### CODE REVIEW

**Comment by DRMacIver** [comment]
One thing to consider here is that there are a number of forks of unittest (e.g. unittest2). We might want to have an extensible mechanism of some sort to list test skipping exceptions.

**Comment by DRMacIver** [comment]
pytest also has its own [skipping mechanism](https://docs.pytest.org/en/latest/skipping.html#imperative-xfail-from-within-a-test-or-setup-function) which presumably has the same problem (less noticeably because pytest won't output on a test failure, but we still probably shouldn't be running the test and shrink process here) and is another argument for an extensible list.

**Comment by alexwlchan** [comment]
Added a prototype of a configurable list. It tries to infer some sensible defaults from the environment, at the cost of several module imports. Does this approach look sensible?

**Comment by DRMacIver** [comment]
The approach looks fairly reasonable. I wonder if reraise\_exceptions would be a better name though? Or something like that. Nothing about what this does is actually about skipping except that that happens to be the example that motivated it.

**Comment by alexwlchan** [comment]
Okay, I think the final version of the core patch is present. I’m trying to write some tests, but a little stymied by the fact that I’m having trouble reproducing the bug within the Hypothesis test suite.

* pytest never seemed to be affected, AFAICT
* I can’t get unittest to spit out the error message within my test
* I haven’t tried to test unittest2 or nose’s behaviour yet

And then extra weird things happen if you run a test with nose but it’s pytest that raises the skip exception… yik.

File: src/hypothesis/core.py [hunk 1]
```diff
@@ -46,6 +46,8 @@
     get_pretty_function_description
 from hypothesis.searchstrategy.strategies import SearchStrategy
 
+import unittest
+
```
Comment (Zac-HD) [inline src/hypothesis/core.py]: Looks like an unused import?  And another at L449.

**Comment by DRMacIver** [comment]
> I can’t get unittest to spit out the error message within my test

I think with all of these it'd be fairly reasonable to write tests that don't try to invoke the test function with the runner: Just call the function/method directly, assert that you caught the right exception, use capture\_out and assert that "Falsifying example" does not appear in the test output.

You'll need to switch to the default reporter as well so that pytest doesn't eat your output.

So basically something like https://github.com/HypothesisWorks/hypothesis-python/blob/master/tests/cover/test_arbitrary_data.py#L43

**Comment by DRMacIver** [comment]
> pytest never seemed to be affected, AFAICT

Actually, here's a suggested test that's more independent of the manifestation and would have shown up under pytest: The reason pytest din't seem to be affected was that it was masking the problem.

Because the real problem is that Hypothesis is treating this as a failure and minimizing it rather than failing fast. So the test should also assert that the function under test is only called once if it raises a skip test immediately.

**Comment by DRMacIver** [comment]
From #535:

> I’m adding reraise_exceptions in #515 as well.

So, thinking about it, I wonder if this actually makes sense in a setting. Can you think of a use case where this should actually be localized to specific tests? I feel like we might be better off just making it a giant global list with an API to add to it.

**Comment by alexwlchan** [comment]
> Can you think of a use case where this should actually be localized to specific tests? I feel like we might be better off just making it a giant global list with an API to add to it.

No.

TBQH, I’m still a bit unsure as to when you’d ever use this _except_ for raising a "skip this test" exception.

**Comment by DRMacIver** [comment]
> No.

In that case I propose that rather than it being a setting we have (internally) a big list of exceptions to ignore and a public API for adding to that list globally.

> TBQH, I’m still a bit unsure as to when you’d ever use this except for raising a "skip this test" exception.

Me too.

**Comment by alexwlchan** [comment]
> I propose that rather than it being a setting we have (internally) a big list of exceptions to ignore and a public API for adding to that list globally.

Agreed, but further: since “public API” = “documented”, and neither of us can think of a reason to do anything besides sensible defaults (at least, for now), I’d propose leaving this private until we have a good reason not to.

**Comment by DRMacIver** [comment]
>  I’d propose leaving this private until we have a good reason not to.

👍 WFM.

**Comment by alexwlchan** [comment]
Okay, tidied and rebased, ready for review.

I think coverage might get upset by this because we only go down one half of those `try … ImportError` blocks.

**Comment by DRMacIver** [comment]
> I think coverage might get upset by this because we only go down one half of those try … ImportError blocks.

I've no problem with those having `# pragma nocover`

**Comment by DRMacIver** [comment]
Huh. https://travis-ci.org/HypothesisWorks/hypothesis-python/jobs/224895255#L361 is an interesting error.

**Comment by alexwlchan** [comment]
Oops, turns out I’d imported the SkipTest plugin, not the exception. Anyway, rebased against current master, list is cached, and now with what I think is the correct set of exceptions to catch.

File: src/hypothesis/core.py [hunk 2]
```diff
@@ -429,6 +429,40 @@ def process_arguments_to_given(
...
+    try:
+        from nose import SkipTest as NoseSkipTest
+        exceptions.append(NoseSkipTest)
+    except ImportError:
+        pass
+
+    return tuple(exceptions)
+
+
+exceptions_to_reraise = skip_exceptions_to_reraise()
```
Comment (DRMacIver) [inline src/hypothesis/core.py]: Nit: this should probably be uppercase

**Comment by Zac-HD** [comment]
@alexwlchan, mind if I rebase this on master so it can be merged?

**Comment by DRMacIver** [comment]
> mind if I rebase this on master so it can be merged?

I think the build here is legitimately broken, so it needs more than a rebase to be mergeable.

**Comment by alexwlchan** [comment]
I’ll do it. Sorry – been busy the last few weeks.

**Comment by DRMacIver** [comment]
> I’ll do it. Sorry – been busy the last few weeks.

No worries. Me too (I've got a few neglected PRs myself).

**Comment by Zac-HD** [comment]
No worries here either, just checking in 😄

**Comment by Zac-HD** [comment]
Looking at the test output, you'll need to add `nose` and `unittest2` to `.isort.cfg`, and I think either add them to the test requirements or make the tests pass when they can't be imported.
